### PR TITLE
Switch between header/source file

### PR DESCRIPTION
**Pull request by Nebiroth**
For now this only switches between header and source files that have the same base name. Once the indexer for clangd is complete we will be using it instead to offer complete coverage. Only covers .cpp and .h files but other extensions will be supported as well.

### CODE REVIEW

**Comment by Nebiroth** [comment]
Forgot to mention that I still need to fix the menu contribution being enabled on files that aren't C/C++ files.

File: src/application/common/commands-common.ts [hunk 1]
```diff
@@ -20,6 +20,7 @@ export namespace CommonCommands {
 
     export const EDIT_UNDO = 'edit_undo';
     export const EDIT_REDO = 'edit_redo';
+    export const SWITCH_HEADER_SOURCE = 'switch_header_source';
```
Comment (akosyakov) [inline src/application/common/commands-common.ts]: It should not appear in the application extension, it is cpp specific and should belong to the cpp extension

File: src/cpp/browser/cpp-protocol.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    severity: MessageType;
+    message: string;
+    data?: any;
+    commands?: Command[];
+}
+
+
+
+export namespace TextDocumentItemRequest {
+    export const type = new RequestType<TextDocumentIdentifier, string | undefined, void, void>("textDocument/sourceHeaderSwitch");
```
Comment (akosyakov) [inline src/cpp/browser/cpp-protocol.ts]: it is up to you, but request names are similar to method names and usually start with a verb, i.e `textDocument/switchSourceHeader`

File: src/filesystem/browser/filesystem-commands.ts [hunk 3]
```diff
@@ -60,7 +66,8 @@ export class FileCommandContribution implements CommandContribution {
         @inject(FileSystem) protected readonly fileSystem: FileSystem,
         @inject(ClipboardService) protected readonly clipboardService: ClipboardService,
         @inject(SelectionService) protected readonly selectionService: SelectionService,
-        @inject(OpenerService) protected readonly openerService: OpenerService
+        @inject(OpenerService) protected readonly openerService: OpenerService,
+        @inject(CppClientContribution) protected readonly clientContribution: CppClientContribution
```
Comment (akosyakov) [inline src/filesystem/browser/filesystem-commands.ts]: Should not be here, the cpp extension can depend on the fs extension, not another way around

File: src/java/browser/java-resource.ts [hunk 4]
```diff
@@ -16,7 +16,7 @@ export class JavaResource implements Resource {
 
     constructor(
         public uri: URI,
-        protected readonly clientContribution: JavaClientContribution
+        protected readonly clientContribution: JavaClientContribution,
```
Comment (akosyakov) [inline src/java/browser/java-resource.ts]: please remove a trailing comma here

File: src/filesystem/browser/filesystem-commands.ts [hunk 5]
```diff
@@ -91,7 +98,23 @@ export class FileCommandContribution implements CommandContribution {
...
                 registry.registerHandler(openWithCommand.id,
                     new FileSystemCommandHandler(this.selectionService, uri => {
-                        return opener.open(uri);
+
+                    })
+                );
+            }
+        });
+
+        this.openerService.getOpeners().then(openers => {
```
Comment (akosyakov) [inline src/filesystem/browser/filesystem-commands.ts]: Please roll back this file, it is the common fs extension, not cpp specific. What are you trying to achieve? You should do it in the cpp extension, via a new opener of a resource. Look at the java extension.

File: src/monaco/browser/monaco-command.ts [hunk 6]
```diff
@@ -48,6 +48,17 @@ export class MonacoEditorCommandHandlers implements CommandContribution {
             commands.registerHandler(id, handler);
         });
 
+        [CommonCommands.SWITCH_HEADER_SOURCE].forEach(id => {
```
Comment (akosyakov) [inline src/monaco/browser/monaco-command.ts]: the same here, should not be here, the cpp extension can use the monaco extenion but not another way around

File: src/monaco/browser/monaco-menu.ts [hunk 7]
```diff
@@ -22,6 +22,9 @@ export class MonacoEditorMenuContribution implements MenuContribution {
         registry.registerMenuAction([EDITOR_CONTEXT_MENU_ID, "1_undo/redo"], {
             commandId: CommonCommands.EDIT_REDO
         });
+        registry.registerMenuAction([EDITOR_CONTEXT_MENU_ID, "1_undo/redo"], {
```
Comment (akosyakov) [inline src/monaco/browser/monaco-menu.ts]: the same here, move it to the cpp extension

File: packages/core/src/common/commands-common.ts [hunk 8]
```diff
@@ -20,6 +20,7 @@ export namespace CommonCommands {
 
     export const EDIT_UNDO = 'edit_undo';
     export const EDIT_REDO = 'edit_redo';
+    export const SWITCH_SOURCE_HEADER = 'switch_header_source';
```
Comment (akosyakov) [inline packages/core/src/common/commands-common.ts]: This file should not be changed in the final PR

File: packages/filesystem/src/browser/filesystem-commands.ts [hunk 9]
```diff
@@ -60,7 +61,7 @@ export class FileCommandContribution implements CommandContribution {
         @inject(FileSystem) protected readonly fileSystem: FileSystem,
         @inject(ClipboardService) protected readonly clipboardService: ClipboardService,
         @inject(SelectionService) protected readonly selectionService: SelectionService,
-        @inject(OpenerService) protected readonly openerService: OpenerService
+        @inject(OpenerService) protected readonly openerService: OpenerService,
```
Comment (akosyakov) [inline packages/filesystem/src/browser/filesystem-commands.ts]: This file should not be changed in the final PR

File: packages/editor/src/browser/editor-command.ts [hunk 10]
```diff
@@ -15,6 +15,9 @@ import { EditorManager } from "./editor-manager";
 export const SHOW_REFERENCES: Command = {
     id: 'textEditor.commands.showReferences'
 };
+export const TEST_COMMAND: Command = {
```
Comment (akosyakov) [inline packages/editor/src/browser/editor-command.ts]: This file should not be changed in the final PR

File: packages/cpp/src/browser/cpp-commands.ts [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+            commandId: SWITCH_SOURCE_HEADER.id
+        });
+    }
+
+    protected switchSourceHeader(): void {
+        const docIdentifier = TextDocumentIdentifier.create(this.selectionService.selection.uri.toString());
+        this.clientContribution.languageClient.then((languageClient) => {
+            languageClient.sendRequest(TextDocumentItemRequest.type, docIdentifier).then(sourceUri => {
+                if (sourceUri !== undefined) {
+                    open(this.openerService, new URI(sourceUri.toString()));
```
Comment (akosyakov) [inline packages/cpp/src/browser/cpp-commands.ts]: Have you managed to get the expected behavior? If not, could you describe it? I can help to figure out how to use `openerService` properly, maybe you need to implement your `OpenHandler` or we need to improve `OpenerService` to support your use case.

Comment (Nebiroth) [inline packages/cpp/src/browser/cpp-commands.ts]: It seems like it's behaving as expected right now. That said, I will probably change the behavior once the clangd indexer is ready to be used.

**Comment by akosyakov** [comment]
@Nebiroth there are still changes outside of `cpp` extension, could you remove them? and then we can merge it

File: packages/cpp/extension.package.json [hunk 12]
```diff
@@ -4,7 +4,10 @@
   "description": "Theia - Cpp Extension",
   "dependencies": {
     "@theia/core": "^0.1.1",
-    "@theia/languages": "^0.1.1"
+    "@theia/languages": "^0.1.1",
+    "@theia/editor": "^0.1.0",
```
Comment (akosyakov) [inline packages/cpp/extension.package.json]: versions should be 0.1.1 for `@theia/*` packages

**Pull request merged**